michbl/assignment2/BinaryTree.hpp

```diff
@@ -84,111 +82,99 @@ class BinaryTree {
+            return true;
         }
-        if (found) {
+
+        if (printAncestors(node->left, key) || printAncestors(node->right, key)) {
             if (node == this->root) {
-                cout << node->key << endl;
+                cout << node->key;
             }
             else {
                 cout << node->key << ", ";
             }
         }
     }
 
-    bool bothKeysAreInSubtree(Node<T> *node, T key1, T key2) {
-        bool found1 = false, found2 = false;
-        bothKeysAreInSubtreeAux(node->left, key1, key2, found1, found2);
-        bothKeysAreInSubtreeAux(node->right, key1, key2, found1, found2);
-        return found1 && found2;
+    bool isAncestor(const Node<T> *node, const T &key1, const T &key2) {
+        bool found1 = keyIsInSubtree(node->left, key1) || keyIsInSubtree(node->right, key1);
+        bool found2 = keyIsInSubtree(node->left, key2) || keyIsInSubtree(node->right, key2);
+        return  found1 && found2;
     }
 
-    void bothKeysAreInSubtreeAux(Node<T> *node, T key1, T key2, bool &found1, bool &found2) {
+    bool keyIsInSubtree(const Node<T> *node, T key) {
         if (node == nullptr) {
-            return;
-        }
-        if (node->key == key1) {
-            found1 = true;
+            return false;
         }
-        if (node->key == key2) {
-            found2 = true;
+        if (node->key == key) {
+            return true;
         }
-        bothKeysAreInSubtreeAux(node->left, key1, key2, found1, found2);
-        bothKeysAreInSubtreeAux(node->right, key1, key2, found1, found2);
+        return keyIsInSubtree(node->left, key) || keyIsInSubtree(node->right, key);
     }
 
-    void printLowestCommonAncestor(Node<T> *node, T key1, T key2, bool &found) {
+    // The return value denotes weather the LCA was already found or not
+    bool printLowestCommonAncestor(const Node<T> *node, const T &key1, const T &key2) {
         if (node == nullptr) {
-            return;
-        }
-        if (!found) {
-            printLowestCommonAncestor(node->left, key1, key2, found);
-        }
-        if (!found) {
-            printLowestCommonAncestor(node->right, key1, key2, found);
-        }
-        if (!found && bothKeysAreInSubtree(node, key1, key2)) {
-            found = true;
-            cout << node->key << endl;
+            return false;
+        }
+        bool foundInLeft = printLowestCommonAncestor(node->left, key1, key2);
+        bool foundInRight = printLowestCommonAncestor(node->right, key1, key2);
+        if (!foundInLeft && !foundInRight) {
+            if (isAncestor(node, key1, key2)) {
+                cout << node->key;
+                return true;
+            }
+            else {
+                return false;
+            }
         }
+        return (foundInLeft || foundInRight);
     }
 
 
 public:
     BinaryTree() {
         this->root = nullptr;
     }
-    BinaryTree(Node<T> *root) {
+    BinaryTree(const BinaryTree &tree) {

```

Comment:
Now `if (root == nullptr)` is going to check the local variable `root`.
Did you mean `tree`?
I how you did not because `tree` is a reference and can never be null :)
In that case, `this->root = copyTree(tree.root);` is enough for this function, because `copyTree` already handles the case in which `tree.root` is null, so no need to check this explicitly :)